src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -71,6 +71,42 @@ public List<TElement> ToList()
             return list;
         }
 
+        public HashSet<TElement> ToHashSet(IEqualityComparer<TElement> comparer) => new HashSet<TElement>(_source, comparer);
+
+        public HashSet<TElement> ToHashSet(int minIdx, int maxIdx, IEqualityComparer<TElement> comparer)
+        {
+            Buffer<TElement> buffer = new Buffer<TElement>(_source);
+
+            int count = buffer._count;
+
+            if (count <= minIdx)
+            {
+                return new HashSet<TElement>(comparer);
+            }
+
+            if (count <= maxIdx)
+            {
+                maxIdx = count - 1;
+            }
+
+            if (minIdx == maxIdx)
+            {
+                return new HashSet<TElement>(1, comparer) { GetEnumerableSorter().ElementAt(buffer._items, count, minIdx) };
+            }
+
+            int[] map = SortedMap(buffer, minIdx, maxIdx);
```

Comment:
The partial quicksort used will sort everything between the indices without bothering to sort those things outside of them in O(n + k log k) time where n is the total number of elements and k is the number of elements in the range sought (in best and average case, it has O(n²) worse case behaviour).
Really all we need here is a double quick partition so that we have the desired elements in the range but don't do the further effort of sorting them, which I think should be possible in O(n) time for best and average case. That would be enough of an improvement to be worth doing, though quite a bit of work.
I think it would be reasonable to take this as an improvement and consider improving further to just partitioning later.